### PR TITLE
prometheus: allow stats to be opt out from automatic namespacing

### LINKED ISSUE #7957: Prometheus: allow opting out of automatic namspacing
*Title*: *Prometheus: allow opting out of automatic namespacing*

*Description*:
PR #2140 added automatic namspacing to all prometheus metrics emitted by envoy per prometheus best practices.

A constant string `envoy_` is prepended to the prometheus metric names, with no way to override or opt out of the namespacing.

As filters add more metrics, they may be namespaced under a different prefix. For example istio would like some filter emitted metrics to be namespaced under `istio_` prefix.
 
Options
1.  Add a convention: example: If a metrics name starts with `_` or some other marker, automatic namspacing will be disabled for that metric. 
2.  Add this option to admin config. 
    a. Add regex list to indicate metric names that would opt out of prometheus automatic namespacing.
    b. Get this config to admin.cc. At present prometheus formatter receives no config.

@ggreenway  @jmarantz

### PR DESCRIPTION
**Pull request by lizan**
Commit Message:
From envoyproxy/envoy-wasm#138. Just add the conventional prefix for now until we see some reason to make it configurable.

Additional Description:
Risk Level: Low
Testing: CI
Docs Changes: code level doc (not user facing)
Release Notes: N/A
Fixes #7957

### CODE REVIEW

**Comment by mandarjog** [comment]
@ggreenway istio stats extensions (s) use it to export metrics with metrics names that are arbitrarily namespaces.

For ex ‘Istio_requests_total’ would be ‘envoy_istio_request_total’ without this change.
Istio plugin exports it as _istio_requests_total.

**Comment by mandarjog** [comment]
@ggreenway see https://github.com/envoyproxy/envoy/issues/7957

**Comment by lizan** [comment]
@ggreenway it doesn't affect statsd or other stuff, it allows extensions to register stats that not being automatically namespaced in prometheus stats. more context are in the issue.

**Comment by ggreenway** [comment]
> @ggreenway it doesn't affect statsd or other stuff, it allows extensions to register stats that not being automatically namespaced in prometheus stats. more context are in the issue.

What I mean is that if you have stats starting with `_`, the prefix is stripped off for prometheus, but what if the user configured a different stats sink? Wouldn't the stats be prefixed with `_`? 

In general, I'm not a fan of encoding things like this into strings. It would be better if a flag could be set on the stat when it is created, or possibly create some rules (regex based maybe?) for which stats should be treated differently. @jmarantz is there a way to keep this metadata with a stat, and set it on creation?

**Comment by jmarantz** [comment]
It should be possible (just annoying due to mocks) to use some spare bits in stats flags for things like this if they are deemed important enough. There are quite a few bits left now in padding. See https://github.com/envoyproxy/envoy/blob/556974437aef10c1eb00569967f4720251545634/include/envoy/stats/stats.h#L96

There a 16 flag bits of which we are using 3 now, and 2 of those are only used for Gauges.

**Comment by lizan** [comment]
> What I mean is that if you have stats starting with `_`, the prefix is stripped off for prometheus, but what if the user configured a different stats sink? Wouldn't the stats be prefixed with `_`?

Yes it will be, it will be up to the extension to prefix or not.

> In general, I'm not a fan of encoding things like this into strings. It would be better if a flag could be set on the stat when it is created, or possibly create some rules (regex based maybe?) for which stats should be treated differently. @jmarantz is there a way to keep this metadata with a stat, and set it on creation?

Configurable rules doesn't work well when an extension is loaded dynamically (e.g. wasm).

> It should be possible (just annoying due to mocks) to use some spare bits in stats flags for things like this if they are deemed important enough. There are quite a few bits left now in padding. See

@jmarantz @ggreenway my take was it isn't important enough for a bit there, but if you feel so I can use that.

**Comment by jmarantz** [comment]
I wasn't voting on importance... Was commenting on feasibility :)

**Comment by mandarjog** [comment]
@ggreenway one requirement is that Changing Filter config thru XDS, does not require changing bootstrap, even with bootstrap

1. Add bootstrap static configuration to peel away _ in stats.
    This is api. We can add a stats.prometheus option to stop name spacing based on prefix.
     Cons: lots of new config 
2. Use opencensus / OT which has APIs to do all this.
3. Use this convention that adds least amount of code.
4. Use spare bits per suggestion, but seems unclean.

**Comment by ggreenway** [comment]
Given that your use-case is for a filter to opt it's metrics out of namespacing, what if we added an API so that filters could register a prefix (at filter load time) to opt out of namespacing? So you could prefix all your metrics with `myfiltername_`, register that prefix with the admin handler through some API, and optionally have `myfiltername_` stripped before the metric is emitted? Then we're not hard-coding a convention that may conflict with future changes we need to make. And no new configuration is needed, bootstrap or otherwise.

**Comment by mandarjog** [comment]
@ggreenway are you suggesting that a statsScope opts out of namespacing during score creation?
Last time I checked, this information was not readily available inside the scraper.

But I do like the solution since it does not add any XDS or bootstrap config.
@lizan @jmarantz  ?

**Comment by ggreenway** [comment]
> @ggreenway are you suggesting that a statsScope opts out of namespacing during scope creation?
> Last time I checked, this information was not readily available inside the prometheus scraper.

Not exactly. I'm suggesting that for the stats you want to exclude from namespacing, you come up with a prefix specific to your filter that you prepend on all your stat names, eg `myfilter_`. Then we add an API to the Server or somewhere similar where you can add to a list of prefixes that will not get the `envoy_` prefix added on for prometheus stats. Then the code for generating prometheus stats will look similar to what it is in this PR now, except that instead of checking for `_`, it looks at a list of prefixes for if any of them match.

**Comment by mandarjog** [comment]
> > @ggreenway are you suggesting that a statsScope opts out of namespacing during scope creation?
> > Last time I checked, this information was not readily available inside the prometheus scraper.
> 
> Not exactly. I'm suggesting that for the stats you want to exclude from namespacing, you come up with a prefix specific to your filter that you prepend on all your stat names, eg `myfilter_`. Then we add an API to the Server or somewhere similar where you can add to a list of prefixes that will not get the `envoy_` prefix added on for prometheus stats. Then the code for generating prometheus stats will look similar to what it is in this PR now, except that instead of checking for `_`, it looks at a list of prefixes for if any of them match.

The details of the `prefix_list` is populated will a difference.
But I can see us calling `registerPrometheusPrefixList()` somewhere in the beginning. This will have to be called on a singleton.

@lizan  @jplevyak  wdyt?

**Comment by jplevyak** [comment]
We can call registerPrometheusPrefixList the first time any Wasm module loads.

**Comment by ggreenway** [comment]
> We can call registerPrometheusPrefixList the first time any Wasm module loads.

Yes, that sounds find. If that's tricky to track, you could maybe use something like `absl::call_once`. Also, I'd name the function `registerPrometheusNamespacePrefix`.

File: source/server/admin/prometheus_stats.cc [hunk 1]
```diff
@@ -11,19 +12,15 @@ namespace Server {
...
 
 const std::regex& promRegex() { CONSTRUCT_ON_FIRST_USE(std::regex, "[^a-zA-Z0-9_]"); }
+const std::regex& namespaceRegex() { CONSTRUCT_ON_FIRST_USE(std::regex, "^[a-zA-Z0-9]*$"); }
 
 /**
  * Take a string and sanitize it according to Prometheus conventions.
  */
 std::string sanitizeName(const std::string& name) {
   // The name must match the regex [a-zA-Z_][a-zA-Z0-9_]* as required by
   // prometheus. Refer to https://prometheus.io/docs/concepts/data_model/.
```
Comment (ggreenway) [inline source/server/admin/prometheus_stats.cc:22]: Can you add onto this comment to say that the initial `[a-zA-Z_]` constraint is always satisfied by the namespace/prefix? (I'm assuming that's why you deleted the number-detection chunk below.)

File: source/server/admin/prometheus_stats.cc [hunk 2]
```diff
@@ -11,19 +12,15 @@ namespace Server {
 namespace {
 
 const std::regex& promRegex() { CONSTRUCT_ON_FIRST_USE(std::regex, "[^a-zA-Z0-9_]"); }
+const std::regex& namespaceRegex() { CONSTRUCT_ON_FIRST_USE(std::regex, "^[a-zA-Z0-9]*$"); }
```
[the remark this answers is not in the document]
Comment (ggreenway) [reply source/server/admin/prometheus_stats.cc:15]: Actually, on second thought, I think this regex should be `^[a-zA-Z_][a-zA-Z0-9]*$` (initial non-digit character)

**Comment by lizan** [comment]
@ggreenway ping?

**Pull request merged**